zcash_proofs/src/sapling/verifier.rs

```diff
@@ -33,18 +34,19 @@ impl SaplingVerificationContextInner {
     #[allow(clippy::too_many_arguments)]
     fn check_spend<C>(
         &mut self,
-        cv: jubjub::ExtendedPoint,
+        cv: &ValueCommitment,
         anchor: bls12_381::Scalar,
         nullifier: &[u8; 32],
-        rk: PublicKey,
+        rk: &PublicKey,
         sighash_value: &[u8; 32],
-        spend_auth_sig: Signature,
+        spend_auth_sig: &Signature,
         zkproof: Proof<Bls12>,
         verifier_ctx: &mut C,
-        spend_auth_sig_verifier: impl FnOnce(&mut C, PublicKey, [u8; 64], Signature) -> bool,
+        spend_auth_sig_verifier: impl FnOnce(&mut C, &PublicKey, [u8; 64], &Signature) -> bool,
         proof_verifier: impl FnOnce(&mut C, Proof<Bls12>, [bls12_381::Scalar; 7]) -> bool,
     ) -> bool {
-        if (cv.is_small_order() | rk.0.is_small_order()).into() {
+        // `cv` is not small order by construction.
+        if rk.0.is_small_order().into() {
```

Comment:
This is consensus-critical code, so the PR requires 3 reviews. My intention in this refactor was to encapsulate this consensus check inside the `ValueCommitment` type (which is partly why I added a separate `CommitmentSum` type, to allow the latter to hold the value "zero", which would fail this check if we did `cv` summing with the same type as in `orchard`).

Reply:
As far as I can see, `ValueCommitment::derive` can return a `ValueCommitment` for the zero point, so this is not equivalent. (`ValueCommitment::from_bytes` indeed does the small order check, but we can't tell by local code review whether `cv` was constructed by `derive` or `from_bytes`.)

Reply:
Aside: I wish we hadn't excluded the zero point here. It's ugly and unnecessary.

Reply:
Note that $\mathsf{ValueCommitment^{Sapling}.Output}$ is $\mathcal{J}$, i.e. it doesn't exclude the zero point. The exclusion of small-order points is in the descriptions of the $\mathsf{cv}$ fields of Spend descriptions in section 4.4, and Output descriptions in section 4.5.

Reply:
👍